python/tank/descriptor/io_descriptor/base.py

```diff
@@ -211,93 +212,85 @@ def _find_latest_tag_by_pattern(self, version_numbers, pattern):
-            #       }
-            #   },
-            #   4: {1: {},
-            #       2: {1: {}, 2: {}},
-            #       3: {}
-            #       }
-            #   }
-            # }
-            #
-            current = versions
-            for number in version_split:
-                if number not in current:
-                    current[number] = {}
-                current = current[number]
-
-        # now search for the latest version matching our pattern
-        if not re.match("^v([0-9]+|x)(.([0-9]+|x)){2,}$", pattern):
+        # Validate the pattern: we expect at least v<digit or x>.<digit or x>.<digit or x>
+        # - we can then have more .<digit or x>
+        # - we can then have a -<topic name>, optionally followed by multiple .<alphanumeric or _>
+        if not re.match("^v(\d+|x)(\.(\d+|x)){2,}(-\w+(\.\w+)*)?$", pattern):
             raise TankDescriptorError("Cannot parse version expression '%s'!" % pattern)
-
-        # split our pattern, beware each part is a string (even integers)
-        version_split = re.findall("([0-9]+|x)", pattern)
-        if 'x' in version_split:
-            # check that we don't have an incorrect pattern using x
-            # then a digit, eg. v4.x.2
-            if re.match("^v[0-9\.]+[x\.]+[0-9\.]+$", pattern):
-                raise TankDescriptorError(
-                    "Incorrect version pattern '%s'. "
-                    "There should be no digit after a 'x'." % pattern
-                )
-
-        current = versions
-        version_to_use = None
-        # process each digit in the pattern
-        for version_digit in version_split:
-            if version_digit == 'x':
-                # replace the 'x' by the latest at this level
-                version_digit = max(current.keys(), key=int)
-            version_digit = int(version_digit)
-            if version_digit not in current:
-                # no matches
+        # Split the pattern either with "." or "-", keep the separator in the result
+        # with the grouping in the regexp
+        version_split = re.split("(\.|-)", pattern)
+        num_tokens = len(version_split)
+        # Special case for the first token for which the wildcard is "vx" and not
+        # "x"
+        has_v_prefix = False
+        version_regexp = version_split[0]
+        if version_split[0] == "vx":
+            has_v_prefix = True
+            version_split[0] = "x"
+        # Build a base version string that versions will have to match. The base
+        # version contains all tokens until we find a "x".
+        for i in range(0, num_tokens, 2):
+            token = version_split[i]
+            if token == "x":
+                # if we found a "x", check that all following tokens are "x" as
+                # well, having something like v4.x.2 is illegal
+                # we can't use 'any' here until supporting Python 2.4 is deprecated
+                for token2 in version_split[i+2:num_tokens:2]:
```

Comment:
Added some comments explaining that `tokens` are every other entry, hence the `2` step.